### ISSUE TITLE
Should we handle Largo widget-specific CSS and Wordpress widget-specific CSS in separate files?

### ISSUE THREAD

**Issue by benlk** [INN/Largo#880]
[less/inc/widgets.less](https://github.com/INN/Largo/blob/master/less/inc/widgets.less) is a mite unorganized now. At the risk of duplicating styles, I think we should do one of the following:

- keep generic-widget styles in `widgets.less`, and move Largo-specific widget styles to `widgets-largo.less` and Wordpress default widget styles to `widgets-wordpress.less`,
- Completely fill out the table of contents in `widgets.less` *now*, with all Largo widgets and all Wordpress widgets that we style, so that the ordering is consistent and not a mix of Largo and Wordpress widgets, intermingled
- Remove the numbering on the table of contents, so that widgets can be more easily shuffled around in the future, to preserve any semblance of divide between Largo and Wordpress widgets

The current ToC of widgets.less:

```
  Contents:
    1.0 - Global widget styles
    2.0 - Follow Widget
    3.0 - About Widget
    4.0 - Donate Widget
    5.0 - Featured and Recent Posts Widget
    6.0 - Series Posts Widget
    7.0 - Related Posts Widget
    8.0 - Roster Widget
    9.0 - WordPress Image Widget
```

Between Largo and Wordpress default widgets, a clean install of Largo with default settings (Eye on Ohio) has 24 widgets:

- [x] Wordpress:
    - [x] *in general*
    - [x] Archives
    - [x] Categories
    - [x] Custom Menu
    - [x] Image Widget
    - [x] Search
    - [x] Text
- [x] Largo
    - [x] INN Member Stories
    - [x] Largo About Site
    - [x] Largo Author Bio
    - [x] Largo Donate Widget
    - [x] Largo Explore Related
    - [x] Largo Facebook Widget
    - [x] Largo Featured Posts
    - [x] Largo Follow
    - [x] Largo Footer Featured Posts, **Deprecated**
    - [x] Largo Prev/Next Links
    - [x] Largo Recent Comments
    - [x] Largo Recent Posts
    - [x] Largo Related Posts
    - [x] Largo Sidebar Featured Posts, **Deprecated**
    - [x] Largo Staff Roster Widget
    - [x] Largo Tag List
    - [x] Largo Taxonomy List
    - [x] Largo Twitter Widget

**Comment by aschweigert** [comment]
i would go one further and argue that

less/widgets.less should include only global widget styles + import statements for each individual widget's styles (only where required, hopefully this is not very often) which are each contained in separate widget files in a subfolder: less/inc/widgets/ similar to how inc/widgets.php has the global widget stuff and inc/widgets/ has the specifics.

**Comment by benlk** [comment]
I like that plan. :smiley:

I'll work on this.

**Comment by benlk** [comment]
PR for this is https://github.com/INN/Largo/pull/882

**Comment by rnagle** [comment]
I think we're on the right track here. It might be helpful to think about this in the context of the other components included in Largo and establish a pattern for the organization of less files that applies to everything, adopting some pattern that makes clear what each file is responsible for.

As far as having a separate file for each widget, even if a widget has no custom styles (i.e., it inherits all of its styles from the core Largo style components -- type, margins, paddings, etc.), I don't think it's necessary to have essentially empty files. This feels overly verbose, especially if we're able to hash out a predictable pattern for the entirety of Largo's less files.

**Comment by benlk** [comment]
How about mirroring the layout of the PHP files in the layout of the LESS files? We kind of already have that going, with `less/inc/widgets/` in PR #882:

```
less/
├── editor-style.less
├── inc
│   ├── ads.less
│   ├── archives.less
│   ├── author.less
│   ├── buttons.less
│   ├── clean-read.less
│   ├── comments.less
│   ├── footer.less
│   ├── forms.less
│   ├── header.less
│   ├── homepage.less
│   ├── icons.less
│   ├── layout.less
│   ├── mixins.less
│   ├── navbar.less
│   ├── posts-and-pages.less
│   ├── print.less
│   ├── reset.less
│   ├── responsive-1200px-min.less
│   ├── responsive-767px-max.less
│   ├── responsive-768px-979px.less
│   ├── responsive-utilities.less
│   ├── search.less
│   ├── single.less
│   ├── sticky.less
│   ├── tables.less
│   ├── typography.less
│   ├── variables.less
│   ├── widgets
│   │   └── (trimmed)
│   ├── widgets.less
│   └── wp-responsive-media.less
└── style.less

```

But a lot of stuff that's in `less/inc/` is in the root dir of Largo, like the archive pages.

```
largo-dev/
├── css
├── docs (trimmed)
├── fonts (trimmed)
├── homepages
│   ├── assets
│   │   ├── css
│   │   ├── img
│   │   ├── js
│   │   └── less
│   ├── layouts
│   ├── templates
│   └── zones
├── img
├── inc
│   ├── avatars
│   │   └── js
│   ├── customizer
│   │   ├── css
│   │   └── js
│   ├── picturefill
│   │   ├── inc
│   │   │   └── templates
│   │   └── js
│   │       └── libs
│   ├── widgets
│   └── wp-taxonomy-landing
│       └── functions
├── js (trimmed)
├── lang
├── less (trimmed)
├── lib
│   ├── clean-contact
│   ├── navis-media-credit
│   │   └── js
│   ├── navis-slideshows
│   │   ├── css
│   │   ├── img
│   │   ├── js
│   │   └── vendor
│   │       └── slick
│   └── options-framework
│       ├── css
│       ├── images
│       │   └── colorpicker
│       └── js
├── partials
└── tests (trimmed)
```

**Comment by benlk** [comment]
Notes from talk:

- Adam: Absence of a file is enough documentation. Ryan :+1: 
- Ryan: Mirroring PHP isn't a good way to organize LESS, since functionality of the LESS doesn't really match the organization of the styles.
- Ryan: We should break up the responsive files, and group those with the other rules. Adam :+1: Ben :+1: 
- We should talk about how we write media queries.
- Adam: If it's not a significant impact in terms of file size, let's just put the media queries next to the things.
- Ben: Is there a flag for grunt-contrib-less that joins together the media queries? That's better.
- Ryan: print, variables, layout, reset, icons, those feel like foundational stuff that should be grouped.
- Adam:: What I tried to do was the variables, the reset, the typography, those were supposed to be globals. It's kind of confusing to have homepage.less here, since those should be either somewhere else in `homepages/` or it needs more docs for why it's here.
- Ryan: In terms of general organization and layout of the directory, we should (maybe this is just docs) generally start paring design elements out of largo, to clean things up visually and to make it easier to build child themes.
- Adam: If there's anything generalizable that we learn from the research below, let's add it to the docs repo on coding styles wrt CSS
- Adam: We'll probably redo all the CSS around 0.6

Conclusions:

- We need to do research on different methods of including responsive rules, and see if there are filesize consideration
- We'll merge this and continue with this pattern of breaking things out
   - But remove the blank files
- Later PR: Break up the responsive.less files
- Let's try to remove more than we're adding

**Comment by benlk** [comment]
#882 merged with the changes for widgets and #888 has been opened for the whole Largo refactor, so I'm going to close this.
